src/storage/StorageHttpDownloadHandler.cpp

```diff
@@ -0,0 +1,180 @@
+            ResponseBuilder(downstream_)
+                .status(200, "SSTFile download successfully")
+                .body("SSTFile download successfully")
+                .sendWithEOM();
+        } else {
+            ResponseBuilder(downstream_)
+                .status(404, "SSTFile download failed")
+                .sendWithEOM();
+        }
+    } else {
+        ResponseBuilder(downstream_)
+            .status(404, "HADOOP_HOME not exist")
+            .sendWithEOM();
+    }
+}
+
+
+void StorageHttpDownloadHandler::onUpgrade(UpgradeProtocol) noexcept {
+    // Do nothing
+}
+
+
+void StorageHttpDownloadHandler::requestComplete() noexcept {
+    delete this;
+}
+
+
+void StorageHttpDownloadHandler::onError(ProxygenError error) noexcept {
+    LOG(ERROR) << "Web service StorageHttpDownloadHandler got error: "
+               << proxygen::getErrorString(error);
+}
+
+std::string StorageHttpDownloadHandler::toStr(folly::dynamic& vals) const {
+    std::stringstream ss;
+    for (auto& counter : vals) {
+        auto& val = counter["value"];
+        ss << counter["name"].asString() << "="
+           << val.asString()
+           << "\n";
+    }
+    return ss.str();
+}
+
+bool StorageHttpDownloadHandler::downloadSSTFiles(const std::string& url,
+                                          int port,
+                                          const std::string& path,
+                                          const std::vector<std::string>& parts,
+                                          const std::string& local) {
+    if (isRunning.test_and_set()) {
+        LOG(ERROR) << "Download is not completed";
+        return false;
+    }
+
+    static folly::IOThreadPoolExecutor executor(FLAGS_download_thread_num);
+    std::condition_variable cv;
+    std::mutex lock;
+    int32_t completed{0};
+    std::atomic<bool> successful{true};
+
+    for (auto& part : parts) {
+        auto downloader = [&]() {
+            auto remotePath = folly::stringPrintf("hdfs://%s:%d%s/%d",
+                                                  url.c_str(), port, path.c_str(),
+                                                  atoi(part.c_str()) + 1);
+            auto command = folly::stringPrintf("hdfs dfs -copyToLocal %s %s",
+                                               remotePath.c_str(), local.c_str());
+            LOG(INFO) << "Download SST Files: " << command;
+            auto result = ProcessUtils::runCommand(command.c_str());
+            if (!result.ok()) {
+                LOG(ERROR) << "Failed to download SST Files: " << remotePath;
+                successful = false;
+            }
+            std::unique_lock<std::mutex> uniqueLock(lock);
+            completed++;
+            if (completed == (int32_t)parts.size()) {
+                cv.notify_one();
+            }
+        };
+        executor.add(downloader);
+    }
+
+    std::unique_lock<std::mutex> uniqueLock(lock);
+    cv.wait(uniqueLock);
```

Comment:
Please do a research on condition variable's *spurious wakeup*.

Again, using `Future` and `collectAll` is recommended.

Reply:
fixed